nano/node/ipc.cpp

```diff
@@ -0,0 +1,682 @@
+	virtual void async_write (std::shared_ptr<std::vector<uint8_t>> buffer_a, std::function<void(boost::system::error_code const &, size_t)> callback_a) = 0;
+};
+
+/** Domain and TCP client socket */
+template <typename SOCKET_TYPE, typename ENDPOINT_TYPE>
+class socket_client : public socket_base, public channel
+{
+public:
+	socket_client (boost::asio::io_context & io_ctx_a, ENDPOINT_TYPE endpoint_a) :
+	socket_base (io_ctx_a), endpoint (endpoint_a), socket (io_ctx_a), resolver (io_ctx_a)
+	{
+	}
+
+	void async_resolve (std::string const & host_a, uint16_t port_a, std::function<void(boost::system::error_code const &, boost::asio::ip::tcp::endpoint)> callback_a)
+	{
+		this->timer_start (io_timeout);
+		resolver.async_resolve (boost::asio::ip::tcp::resolver::query (host_a, std::to_string (port_a)), [this, callback_a](boost::system::error_code const & ec, boost::asio::ip::tcp::resolver::iterator endpoint_iterator_a) {
+			this->timer_cancel ();
+			boost::asio::ip::tcp::resolver::iterator end;
+			if (!ec && endpoint_iterator_a != end)
+			{
+				endpoint = *endpoint_iterator_a;
+				callback_a (ec, *endpoint_iterator_a);
+			}
+			else
+			{
+				callback_a (ec, *end);
+			}
+		});
+	}
+
+	void async_connect (std::function<void(boost::system::error_code const &)> callback_a)
+	{
+		this->timer_start (io_timeout);
+		socket.async_connect (endpoint, [this, callback_a](boost::system::error_code const & ec) {
+			this->timer_cancel ();
+			callback_a (ec);
+		});
+	}
+
+	void async_read (std::shared_ptr<std::vector<uint8_t>> buffer_a, size_t size_a, std::function<void(boost::system::error_code const &, size_t)> callback_a) override
+	{
+		this->timer_start (io_timeout);
+		buffer_a->resize (size_a);
+		boost::asio::async_read (socket, boost::asio::buffer (buffer_a->data (), size_a), [this, callback_a](boost::system::error_code const & ec, size_t size_a) {
+			this->timer_cancel ();
+			callback_a (ec, size_a);
+		});
+	}
+
+	void async_write (std::shared_ptr<std::vector<uint8_t>> buffer_a, std::function<void(boost::system::error_code const &, size_t)> callback_a) override
+	{
+		this->timer_start (io_timeout);
+		boost::asio::async_write (socket, boost::asio::buffer (buffer_a->data (), buffer_a->size ()), [this, callback_a, buffer_a](boost::system::error_code const & ec, size_t size_a) {
+			this->timer_cancel ();
+			callback_a (ec, size_a);
+		});
+	}
+
+	/** Shut down and close socket */
+	void close () override
+	{
+		socket.shutdown (boost::asio::ip::tcp::socket::shutdown_both);
+		socket.close ();
+	}
+
+private:
+	ENDPOINT_TYPE endpoint;
+	SOCKET_TYPE socket;
+	boost::asio::ip::tcp::resolver resolver;
+	std::chrono::seconds io_timeout{ 60 };
+};
+
+/**
+ * PIMPL class for ipc_client. This ensures that socket_client and boost details can
+ * stay out of the header file.
+ */
+class client_impl : public nano::ipc::ipc_client_impl
+{
+public:
+	client_impl (boost::asio::io_context & io_ctx_a) :
+	io_ctx (io_ctx_a)
+	{
+	}
+	~client_impl ()
```

Comment:
Empty destructor can be removed

Reply:
Seems like that may fail on some compilers, but let's try and see what CI thinks (https://stackoverflow.com/questions/8619708/must-provide-destructor-in-the-pimpl)